packages/puppeteer-core/src/common/bidi/Realm.ts

```diff
@@ -27,43 +27,60 @@ export const getSourceUrlComment = (url: string): string => {
 
 export class Realm extends EventEmitter {
   readonly connection: Connection;
-  readonly #id: string;
 
+  #id!: string;
   #sandbox!: Sandbox;
 
-  constructor(connection: Connection, id: string) {
+  constructor(connection: Connection) {
     super();
     this.connection = connection;
-    this.#id = id;
   }
 
   get target(): Bidi.Script.Target {
     return {
-      context: this.#id,
+      context: this.#sandbox.environment.context().id,
       sandbox: this.#sandbox.name,
     };
   }
 
+  handleRealmDestroyed = async (
+    params: Bidi.Script.RealmDestroyed['params']
+  ): Promise<void> => {
+    if (params.realm === this.#id) {
+      const promise = this.internalPuppeteerUtil;
+      this.internalPuppeteerUtil = undefined;
+      try {
+        await (await promise)?.dispose();
+      } catch (error) {
+        debugError(error);
+      }
+    }
+  };
+
+  handleRealmCreated = (params: Bidi.Script.RealmCreated['params']): void => {
+    if (
+      params.type === 'window' &&
+      params.context === this.#sandbox.environment._id &&
+      params.sandbox === this.#sandbox.name
+    ) {
+      this.#id = params.realm;
+      void this.#sandbox.taskManager.rerunAll();
+      this.connection.off(
+        Bidi.ChromiumBidi.Script.EventNames.RealmCreated,
+        this.handleRealmCreated
+      );
+      this.connection.on(
+        Bidi.ChromiumBidi.Script.EventNames.RealmDestroyed,
+        this.handleRealmDestroyed
+      );
+    }
+  };
+
   setSandbox(sandbox: Sandbox): void {
     this.#sandbox = sandbox;
-
-    // TODO: Tack correct realm similar to BrowsingContexts
-    this.connection.on(Bidi.ChromiumBidi.Script.EventNames.RealmCreated, () => {
-      void this.#sandbox.taskManager.rerunAll();
-    });
-    // TODO(jrandolf): We should try to find a less brute-force way of doing
-    // this.
     this.connection.on(
```

Comment:
Let use the same thing we do in Page where we have map between the two.
We can subscribe to both events here as the second will be No-Op if the Id is undefined. 
Also we need a way to unsubscribe so that the object has no reference from `connection`.

Reply:
Hmm, but isn't it more precise to subscribe only when the initial subscription is active? I don't see value in keeping it as a map here (it's not like we will have more added).

Reply:
It make thing easy to follow. Subscribe on construction or in this case setting sandbox then unsubscribe on dispose.

Reply:
Subscription is low-cost, but not the effects. It would pollute the listener pool.

Reply:
Disagree we are doing check to validate thing before doing work. 
Deferring to @OrKoN.

Reply:
I think we need to unsubscribe here because the subscription is global and it will keep realms alive. Also, the pattern of saving listeners to make sure the same listeners are disposed sounds much preferred to me.

Reply:
`RealmCreated` and `RealmDestroyed` will always come in pairs, so the clean up is contained within the listeners themselves.